### DISCUSSION TITLE
Send traces/metrics using OTLP from .net 4.8 apps to the dashboard?

### DISCUSSION THREAD

**Discussion by gentledepp** [microsoft/aspire#4842] [Q&A]
Hello!

I am currently trying to start migrating a .net 4.8 MVC app and some windows service to .new core.

There is a quite [good doc](https://learn.microsoft.com/en-us/aspnet/core/migration/inc/start?view=aspnetcore-8.0) on hosting the new asp.net core version alongside to the legacy MVC app and migrate endpoints incrementally.

Now, I would like to use Aspire to 
- help me run all that stuff (windows service, mvc, mvc core, db, etc.)
- and have one common view on the logs/metrics

Using aspire to launch my console app (aka "windows service") and the Aspnet MVC app does work fine.
However, the dashboard shows no data.

I added  the opentelemetry stuff 
```csharp
     private TracerProvider _tracerProvider;
     private MeterProvider _meterProvider;

     protected override void Application_Start(object sender, EventArgs e)
     {
         // set application version at startup
         var version =  typeof(MvcApplication).Assembly.GetCustomAttribute<AssemblyFileVersionAttribute>().Version;

         var useOtlpExporter = !string.IsNullOrWhiteSpace(System.Configuration.ConfigurationManager.AppSettings["OTEL_EXPORTER_OTLP_ENDPOINT"]);
         var tracerBuilder = Sdk.CreateTracerProviderBuilder();

         if (useOtlpExporter)
             tracerBuilder.AddOtlpExporter();

         _tracerProvider = tracerBuilder
             .AddAspNetInstrumentation()
             .AddSource($"web_{Environment.MachineName}")
             .SetResourceBuilder(
                 ResourceBuilder.CreateDefault()
                     .AddService(serviceName: $"web_{Environment.MachineName}", serviceVersion: version))

             .Build();

         var meterBuilder = Sdk.CreateMeterProviderBuilder();
         if (useOtlpExporter)
             meterBuilder.AddOtlpExporter();

         _meterProvider = meterBuilder
             .AddRuntimeInstrumentation()
             .Build();
```

Note: The console app looks similar but does not have the `AspNetInstrumentation()`

Now, use the [configurationbuilder](https://learn.microsoft.com/en-us/aspnet/config-builder) stuff to get environment variables into my `ConfigurationManager.AppSettings` so aspire can forward all relevant configuration options to my service/web app (e.g. `OTEL_EXPORTER_OTLP_ENDPOINT`) and this works as I do get all the variables that are defined.

Unfortunately, I still do not see any logs/traces/metrics in the Aspire Dashboard.

My guess is:
- Aspire Dashboard [only supports the GRPC protocol](https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/dashboard/standalone?tabs=bash#configure-opentelemetry-sdk)
- but GRPC is not supported by .NET 4.8 [Supported client platforms](https://learn.microsoft.com/en-us/aspnet/core/grpc/supported-platforms?view=aspnetcore-8.0#net-grpc-client-requirements)

Am I right?

If so, are there any plans to somehow support .net 4.8 by [implementing any of the other protocols (http/protobuf, http/json)](https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/#otel_exporter_otlp_protocol) in the Aspire Dashboard?

**Comment by KalleOlaviNiemitalo** [comment]
I have a service in development, running on .NET Framework on Windows 10 and exporting telemetry over OTLP to .NET Aspire dashboard version 8.0.0 on `http://localhost:18889` (thus not using TLS).  Logs, traces, and metrics all show up on the dashboard.  I am not using .NET Aspire orchestration; I start the dashboard from the command line, and the service from the command line or as a Visual Studio project.

Wireshark shows the protocol is some sort of GRPC over HTTP/2 over TCP over IPv6 (`::1`).  HTTP headers include:

```
content-type: application/grpc
te: trailers
grpc-accept-encoding: identity, deflate, gzip
grpc-timeout: 9910m
user-agent: grpc-csharp/2.44.0 (.NET Framework 4.8.9256.0; CLR 4.0.30319.42000; net45; x64) grpc-c/22.0.0 (windows; chttp2)
```

This service references the OpenTelemetry.Exporter.OpenTelemetryProtocol 1.8.1 package (not yet upgraded to 1.9.0).  On net462, that package depends on "Grpc (>= 2.44.0 && < 3.0.0)" and thus the [gRPC C# core-library](https://learn.microsoft.com/en-us/aspnet/core/grpc/netstandard?view=aspnetcore-8.0#grpc-c-core-library), which will be deprecated some time after October 2024, rather than the Grpc.Net.Client package.  Related: <https://github.com/open-telemetry/opentelemetry-dotnet/issues/4613>.

Might you have an authentication problem?  I haven't bothered to secure the OTLP endpoint, because it's on localhost and used only for development so far.

**Reply by KalleOlaviNiemitalo** [reply]
Re starting the dashboard from the command line: I actually have a Windows Terminal profile that runs the following PowerShell script.

```PowerShell
[CmdletBinding()]
Param ()

Set-StrictMode -Version:3.0

$dotnet_exe = Get-Command -Name:'dotnet' -CommandType:Application
if ($null -eq $dotnet_exe) {
    exit 1
}

$dashboard_versions_dir = Join-Path -Path:(Split-Path -Path:($dotnet_exe.Path) -Parent) -ChildPath:'packs/Aspire.Dashboard.Sdk.win-x64'

$dashboard_version_dirs = Get-ChildItem -LiteralPath:$dashboard_versions_dir -Directory | Sort-Object -Property:'Name'

$dashboard_exe = $null
foreach ($dashboard_version_dir in $dashboard_version_dirs) {
    $dashboard_exe_candidate = Join-Path -Path:($dashboard_version_dir.PSPath) -ChildPath:'tools/Aspire.Dashboard.exe'
    if (Test-Path -LiteralPath:$dashboard_exe_candidate) {
        $dashboard_exe = $dashboard_exe_candidate
    }
}

if ($null -eq $dashboard_exe) {
    throw 'Cannot find Aspire.Dashboard.exe'
}

# .NET Aspire 9.0.0-preview.3.24210.17
# https://github.com/dotnet/aspire/blob/c2d83148f169bdf4980393a66eff2a28e94eb6c9/src/Aspire.Dashboard/DashboardWebApplication.cs
$Env:ASPNETCORE_Frontend__AuthMode = 'Unsecured'
$Env:ASPNETCORE_URLS = 'http://localhost:18888/'
$Env:ASPNETCORE_Otlp__AuthMode = 'Unsecured'
$Env:DOTNET_DASHBOARD_OTLP_ENDPOINT_URL = 'http://localhost:18889/'

# Start-Process -FilePath:'microsoft-edge:http://localhost:18888/'

Set-Location -LiteralPath:(Split-Path -Path:$dashboard_exe -Parent)
& $dashboard_exe
```

**Reply by gentledepp** [reply]
Thanks for the answer.

Well, orchestration is what I actually want from aspire :-D

You may be right and I may have an authorization issue - I will try that out.
However, I would have guessed that any API key necessary would be silently passed through as an environment variable just as the `OTEL_EXPORTER_<xyz>` stuff.

**Reply by gentledepp** [reply]
well it turns out that my assumption was correct:
the http headers for api key authorization are passed as an environment variable
and are already set to the OtlpExporterOptions object
https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/dashboard/security-considerations?tabs=bash#secure-telemetry-endpoint

![image](https://github.com/user-attachments/assets/809af811-3956-44de-a6b8-eb158a4e1799)

Still - no communication.


by the way, 1.9.0 still depends on the Grpc 2.44.0 library:
![image](https://github.com/user-attachments/assets/954e6999-0d39-47e6-82fa-1b16b4e25699)

**Reply by KalleOlaviNiemitalo** [reply]
Does the same dashboard show any telemetry from services on .NET 8?  Just trying to understand whether the problem is specific to .NET Framework.

**Reply by gentledepp** [reply]
Hmmm.. interesting question. I dod not even try that one yet since I did not do anything "fancy" yet.

But ok:
Adding an empty .net8 MVC App and wiring up the service defaults like:
![image](https://github.com/user-attachments/assets/e3dd240e-03fd-4252-a174-5c7f0e20bf17)


results in:
![image](https://github.com/user-attachments/assets/cdfa1070-82fc-41bd-8dc4-162df539ffcc)

**Reply by KalleOlaviNiemitalo** [reply]
I don't think that last screen shot shows anything that the ASP.NET Core MVC app has exported over OTLP…?  The "Ressourcen" list and the log messages in the "Auswählen" window appear to originate from .NET Aspire orchestration.

**Reply by gentledepp** [reply]
You are right - It seems I misconfigured the empty mvc app.
I corrected this:
```csharp
var builder = WebApplication.CreateBuilder(args);
builder.AddServiceDefaults();
var app = builder.Build();

app.MapGet("/", () => "Hello World!");

app.Run();


```

and now it shows up in the traces:
![image](https://github.com/user-attachments/assets/2c6d0d36-8e42-4e1d-a892-13a0dfd96aff)


but the other 2 apps (.net 4.8) still do not.
However, I did check that they get the authentication header set for OTLP.

So what else can it be?

**Reply by KalleOlaviNiemitalo** [reply]
"Strukturierte Protokolle" is logs, rather than traces or metrics.  But the code you posted doesn't do anything to export logs.

**Reply by KalleOlaviNiemitalo** [reply]
Also… even though AddRuntimeInstrumentation creates a Meter, I'm not sure it's enabled for OTLP exporting by default.

**Reply by gentledepp** [reply]
Oh... so "traces" is "console" then?

**Reply by KalleOlaviNiemitalo** [reply]
<https://github.com/open-telemetry/opentelemetry-dotnet> maps them like so:

- Microsoft.Extensions.Logging.ILogger to OpenTelemetry logs ("Strukturiert")
- System.Diagnostics.Activity to OpenTelemetry traces ("Überwachungen")
- System.Diagnostics.Metrics.Meter to OpenTelemetry metrics ("Metriken")

The .NET Aspire dashboard does not show the "Ressourcen" and "Konsole" tabs when I start it without .NET Aspire orchestration.  I don't think those tabs rely on OTLP.

**Reply by gentledepp** [reply]
hmmm. ok. But now I am confused.
I added the AspNetInstrumentation 
![image](https://github.com/user-attachments/assets/c5965e61-44cc-4441-b521-efdb50a9cfdc)

And thought I would get a similar "logs" experience as with the asp.net core web application:
That it shows any request and subsequent calls in the "strukturierte protokolle" view :-|

**Reply by KalleOlaviNiemitalo** [reply]
According to <https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/Instrumentation.AspNet-1.9.0-beta.1/src/OpenTelemetry.Instrumentation.AspNet/README.md>, OpenTelemetry.Instrumentation.AspNet supports metrics and traces but not logs.

Did you add the TelemetryHttpModule to Web.config as documented there?

**Reply by KalleOlaviNiemitalo** [reply]
In the "Strukturierte Protokolle" tab whose screen shot you posted in <https://github.com/dotnet/aspire/discussions/4842#discussioncomment-10051643>, what options are there in the dropdown list above the table, other than "(Alle)"?  That shows the resources that have been exported over OTLP, even if no traces/logs/metrics have been exported for those resources yet.  So if your `$"web_{Environment.MachineName}"` shows up there, then that indicates the OTLP connection is OK.

**Reply by gentledepp** [reply]
wait... I'll create a git repo

**Reply by gentledepp** [reply]
Hi! So I created a repro sample here
https://github.com/gentledepp/POC_AspireForNetFramework

I would hope that I get the timeline view for the traces as described here:
https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/dashboard/explore#trace-examples
![image](https://github.com/user-attachments/assets/092d4c7e-3e5a-417f-8cca-1cde5f7464e9)

In the end, I would like to use the strangler fig pattern to move from my old legacy mvc app to mvc core step by step.
And doing so I would like to see the traces for anything that happens between those two web servers :-|

**Reply by gentledepp** [reply]
Two interesting findings:
1) The aspnetintegration seems to be working as the breakpoint in the filter for traced requests is being hit:
![image](https://github.com/user-attachments/assets/11055d1a-d7ab-47b6-a3b2-aba477359195)
2) Interestingly, the Aspire Dashboard does not show any endpoint for the legacy Asp.Net application:
![image](https://github.com/user-attachments/assets/b28295ee-d9c8-4cf5-a410-3f1367b862fa)

**Reply by KalleOlaviNiemitalo** [reply]
I didn't try your repro but I set up something similar.  The problem seems to be specific to `https`. If I run the AppHost project with a launch profile that sets environment variables `"DOTNET_DASHBOARD_OTLP_ENDPOINT_URL": "http://localhost:19161"` and `"ASPIRE_ALLOW_UNSECURED_TRANSPORT": "true"`, then I get traces and metrics from the .NET Framework ASP.NET MVC application all right.

The `aspnetmvc` resource name in the following screen shots comes from the `builder.AddProject<Projects.WebApplication1>("aspnetmvc")` call in the AppHost project.  It goes to an environment variable that `ResourceBuilder.CreateDefault()` picks up.

![screen shot of NET Aspire dashboard showing traces from .NET Framework ASP.NET MVC](https://github.com/user-attachments/assets/44a98f5b-7cc3-4361-92d8-23d40deee809)

![screen shot of .NET Aspire dashboard showing an error trace from .NET Framework ASP.NET MVC](https://github.com/user-attachments/assets/32aca503-c252-40a2-9495-ce2c93d0bdca)

![screen shot of .NET Aspire dashboard showing metrics from .NET Framework ASP.NET MVC](https://github.com/user-attachments/assets/7a8bb24e-5bbc-4efe-93e5-24077a95e75d)

I hope there is some way to make it work with `https` in production environments.  If not directly, then perhaps via the OpenTelemetry Collector.

[OpenTelemetry SDK self-diagnostics](https://github.com/open-telemetry/opentelemetry-dotnet/blob/core-1.9.0/src/OpenTelemetry/README.md#self-diagnostics) seem somewhat usable for logging the exceptions thrown from the OTLP exporter, without breaking to Visual Studio and causing gRPC deadline errors.

The `GET {controller}/{action}/{id}` span name matches [semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/docs/http/http-spans.md#name) but it'll be pretty useless if every span has that name.  It can be overridden by registering these in AspNetTraceInstrumentationOptions:

```csharp
        private static void EnrichWithHttpRequest(Activity activity, HttpRequest httpRequest)
        {
            // Don't set Activity.DisplayName here, because:
            //
            // * ASP.NET hasn't parsed the route values from the request yet.
            // * OpenTelemetry.Instrumentation.AspNet will overwrite DisplayName
            //   when the request finishes.
            //   <https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1948>
            //
            // Just save the HttpRequest so that EnrichWithHttpResponse can find it.
            activity.SetCustomProperty("HttpRequest", httpRequest);
        }

        private static void EnrichWithHttpResponse(Activity activity, HttpResponse response)
        {
            // Get the HttpRequest that was saved by EnrichWithHttpRequest.
            if (activity.GetCustomProperty("HttpRequest") is HttpRequest httpRequest)
            {
                RouteData routeData = httpRequest.RequestContext.RouteData;
                if (routeData.Route is Route route
                    && route.Url == "{controller}/{action}/{id}"
                    && routeData.Values.TryGetValue("controller", out object controller)
                    && routeData.Values.TryGetValue("action", out object action))
                {
                    // Replace the "GET {controller}/{action}/{id}" or similar
                    // DisplayName with something that shows the controller and action.
                    // Omit any "id" route value though; too high cardinality.
                    activity.DisplayName = $"{httpRequest.HttpMethod} {controller}/{action}/{{id}}";
#if false // not part of OpenTelemetry HTTP semantic conventions
                    activity.AddTag("http.route_data.controller", controller);
                    activity.AddTag("http.route_data.action", action);
#endif
                }
            }
        }
```

**Reply by gentledepp** [reply]
can you please provide the full code so I can check?

**Reply by KalleOlaviNiemitalo** [reply]
[AspireApp1.zip](https://github.com/user-attachments/files/16540446/AspireApp1.zip)

**Reply by gentledepp** [reply]
Thank you for the sample!
It works now.

Interesting that "HTTPS" seems to be the culprit here. @davidfowl  could you be so kind and forward this to somebody that could explain this or help out? (Sorry for bothering you :-|)

**Reply by CZEMacLeod** [reply]
I have some stuff to help with this in an example I set up at [C3D.Extensions.Aspire.IISExpress](https://github.com/CZEMacLeod/C3D.Extensions.Aspire.IISExpress) and there is some discussion in #4984.
As you say - one of the big issues seems to be the otlp/grpc implementation for net framework doesn't seem to support https.

**Reply by patiqs** [reply]
> Interesting that "HTTPS" seems to be the culprit here. 

Is there a plan to support HTTPS for GRPC in .NET Framework? I have the same problem described here: https://github.com/dotnet/aspire/issues/4984#issuecomment-2559132973
